src/auth.py

```diff
@@ -27,9 +27,10 @@ class Acl:
 class KafkaAuth:
     """Object for updating Kafka users and ACLs."""
 
-    def __init__(self, opts: List[str], zookeeper: str):
+    def __init__(self, opts: List[str], zookeeper: str, ssl=False):
         self.opts = opts
         self.zookeeper = zookeeper
+        self.ssl = ssl
```

Comment:
What was the reason for not routing this through `charm`?

I was thinking something like referencing `self.charm.tls.enabled` or `self.charm.zookeeper_config.is_tls` (not sure that exists but you get the point).

I think there's an advantage to not having it be declared during the initialization of any 'KafkaAuth' instance, but instead managed through state from the `KafkaTLS` object, as we can separate how we decide 'is using tls' there.

Reply:
Upon further review, I think most of the issues could be solved by adding to `charm.py` events a `if not self.cluster.relation: event.defer() return`

Reply:
Yeah, I can add `charm` as part of the class. I wasn't using it mainly not to change too much of the class.